filesystem.go

```diff
@@ -135,10 +134,29 @@ func (g *FS) Join(elem ...string) string {
 	return filepath.Join(elem...)
 }
 
-// ReadDir is not implemented by the underlying library. Added so billy.Dir
-// is implemented as it is needed by tests.
+// ReadDir implements billy.Dir interface.
 func (g *FS) ReadDir(path string) ([]os.FileInfo, error) {
-	return nil, fmt.Errorf("ReadDir not implemented")
+	d, err := g.v.Open(path)
+	if err != nil {
+		return nil, err
+	}
+
+	defer d.Close()
+
+	files, err := d.Readdir(0)
+	if err != nil {
+		return nil, err
+	}
+
+	// gluster Readdir returns also "." and ".."
+	clean := make([]os.FileInfo, 0, len(files))
+	for _, f := range files {
```

Comment:
maybe instead of copy objects to the new slice, we can find indexes with "." and/or ".." and at the end return the reference to the shifted slice 
for instance:
```go
        n := len(files)
	for i := 0; i < n; i++ {
		if files[i].Name() != "." && files[i].Name() != ".." {
                       // swap with the last element
			files[i], files[n-1] = files[n-1], files[i]
			n--
			i--
		}
	}
       return files[0:n]
```
WDYT? Maybe it sounds like a micro optimization but if you have many files, it can avoid lot of reallocations.

Reply:
Added the optimization and an extra check to test empty directories. Also fixed a bug deleting directories.